x/valset-pref/README.md

```diff
@@ -153,3 +153,41 @@ The Code Layout is very similar to TWAP module.
 - keeper.go - generic SDK boilerplate (defining a wrapper for store keys + params)
 - msg_server.go - handle messages request from client and process responses. 
 - store.go - Managing logic for getting and setting things to underlying stores (KVStore)
+
+## Redelegate algorithm logic pseudocode
+
+Existing ValSet   20osmos {ValA-> 0.5, ValB-> 0.3, ValC-> 0.2} [ValA-> 10osmo, ValB-> 6osmo, ValC-> 4osmo]
+New ValSet        20osmos {ValD-> 0.2, ValE-> 0.2, ValF-> 0.6} [ValD-> 4osmo, ValE-> 4osmo, ValD-> 12osmo]
+
+- // Rearranging the exisintValSet and newValSet to to add extra validator padding
+  - existing_valset_updated = [ValA: 10, ValB: 6, ValC: 4, ValD: 0, ValE: 0, ValF: 0]
+  - new_valset_updated = [ValD: 4, ValE: 4, ValF: 12, ValA: 0, ValB: 0, ValC: 0]
+
+  // calculate the difference between two sets
+  - diff_arr = [ValA: 10, ValB: 6, ValC: 4, ValD: -4, ValE: -4, ValF: -12]
+
+	// Algorithm starts here
+- for i, validator in diff_arr: 
+  - if validator.amount > 0 
+    - for validator.amount > 0: 
+        source_validator = validator.address
+        target_validator = FindMin(diff_arr)  
+
+        amount_to_redelegate = FindMin(target_validator.amount, validator.amount)
+        sdk.BeginRedelegation(delegator, source_validator, target_validator, amount_to_redelegate) 
+
+        diff_arr[i].amount = target_validator.amount + amount_to_redelegate 
+        validator.amount = validator.amount - amount_to_redelegate
+
+- Result 
+  1. diff_arr = [ValA: 0, ValB: 0, ValC: 0, ValD: 0, ValE: 0, ValF: 0]
+  2. [ValA: 0, ValB: 0, ValC: 0, ValD: 4, ValE: 4, ValF: 12] // final result
+
+
+## Redelegation Constraints 
+1. ValA -> ValB redelegate upto 7 times in 21 day period 
+2. ValA -> ValB (redelegate) ValB -> ValC (redelegate) CONSECUTIVE REDELEGATION DOES NOT WORK 
+3. Once you redelegate from ValA -> ValB, you will not be able to redelegate from ValB to another validator for the next 21 days.
+  - the validator on the receiving end of redelegation will be on a 21-day redelegation lock
+
```

Comment:
I thought the constraint was seven times, period. Not seven times over a 21 day period?

Reply:
So i found a [medium](https://medium.com/node-a-team/cosmos-hub-security-vulnerability-regarding-redelegation-and-unbonding-e6a2cf03e928) article thats says "7 times in 21 days"  and I got this from the cosmo-sdk github

From cosmos-sdk [github](https://github.com/cosmos/cosmos-sdk/tree/main/x/staking#redelegation) 
To prevent "redelegation hopping" redelegations may not occur under the situation that:
1. the (re)delegator already has another immature redelegation in progress with a destination to a validator (let's call it Validator X)
2. 
3. and, the (re)delegator is attempting to create a new redelegation where the source validator for this new redelegation is Validator X.